### PR TITLE
MultiText Field

### LINKED ISSUE #3815: Text Field with PredefinedList editor: Select multiple values
Would it be possible to allow to select multiple values in a TextField with PredefinedList?

The editor would allow to choose an EditorOption like `Checkboxes`, `Listbox` or a multi select component

/cc @Skrypt

### PR DESCRIPTION
**Pull request by agriffard**
Field with multiple selectable values.
Similar to the PredefinedList of the TextField but stores the Values in a string[].

Fixes #3815

Questions:
- Is the name correct? `MultiText`
- What editors do we use? Currently with Listbox and Checkboxes
- Do we split in different Field editors?
- Can we select multiple default values?
- How do we display the result? string.Join with ', '?
- Can we index and how?
- Can we search in SQL?
- Is there a specific GraphQL index?
- Do we implement the Required attribute?

### CODE REVIEW

**Comment by hishamco** [comment]
> Is the name correct? MultiValue

Good, but `MultiValuesTextField` may better, because `ListBox`, `Checbox` are also MultiValue fields

> Can we select multiple default values?

It's good for some cases, for instance job skillset requirements

> How do we display the result? string.Join with ', '?

We can have a `Separator` attribute

> Can we search in SQL?

May be

> Do we implement the Required attribute?

Yep

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/MultiSelectFieldIndexHandler.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+{
+    public class MultiSelectFieldIndexHandler : ContentFieldIndexHandler<MultiSelectField>
+    {
+        public override Task BuildIndexAsync(MultiSelectField field, BuildFieldIndexContext context)
+        {
+            var options = context.Settings.ToOptions();
+
+            foreach (var key in context.Keys)
+            {
+                context.DocumentIndex.Set(key, string.Join(",", field.Values), options);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/MultiSelectFieldIndexHandler.cs]: This should be multiple entries on the same field instead. So you can search for a specific entry.

**Comment by sebastienros** [comment]
demo and discussions tomorrow?

**Comment by peterkeating** [comment]
@agriffard Looking really good. Small design tweak, could the `select` used in the standard editor not have a [`min-height` applied to it](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore.Themes/TheAdmin/Assets/scss/TheAdmin.scss#L1013)? When there aren't many values it takes up a lot of unnecessary space. Not sure if that `min-height` is catering for something else in the admin area.

**Comment by MarcBruins** [comment]
Any update on this? is there something still blocking this from getting merged, is so, can i help?

**Comment by deanmarcussen** [comment]
So what does this need @agriffard ?

Looks like it needs lucene indexes fixing, graphql, sql field indexing?

**Comment by agriffard** [comment]
> Looks like it needs lucene indexes fixing, graphql, sql field indexing?

Anyone wants to help finalizing this?

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/MultiSelectFieldIndexHandler.cs (hunk 1 above)
Comment (deanmarcussen) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/MultiSelectFieldIndexHandler.cs:15]: Done

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/MultiSelectFieldSettings.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using Newtonsoft.Json;
+
+namespace OrchardCore.ContentFields.Settings
+{
+    public class MultiSelectFieldSettings
+    {
+        public string Hint { get; set; }
+        public bool Required { get; set; }
+        public string Options { get; set; }
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/MultiSelectFieldSettings.cs:9]: This should just be a `List<MultiSelectListValueOption`. Deserialized once in the driver, and stored on the settings.

You may need a ViewModel

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/MultiSelectFieldSettingsDriver.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                var model = new MultiSelectSettingsViewModel();
+                var settings = new MultiSelectFieldSettings();
+
+                await context.Updater.TryUpdateModelAsync(model, Prefix);
+
+                try
+                {
+                    settings.Required = model.Required;
+                    settings.Hint = model.Hint;
+                    settings.Options = model.Options; // string.IsNullOrWhiteSpace(model.Options)
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Settings/MultiSelectFieldSettingsDriver.cs:45]: Yeah here, deserialize it once, store it as a list. (stops all the `JsonConvert` all over the place

**Review by deanmarcussen — COMMENTED** [review]
Ok @agriffard with you again now for demos.

I added SQL Field indexes, Graphql, and tweaked the lucene, and a few null checks where it was going awry.

Personally I think the two editor types would be much better just combined into one using the vuejsmultiselect. Simple, easier, and a better ui for both.

**Comment by deanmarcussen** [comment]
I'd probably call it a `MultiText` field as well, rather than `MultiSelect`

**Comment by agriffard** [comment]
> I'd probably call it a `MultiText` field as well, rather than `MultiSelect`

It was called MultiValue Field at first.

**Comment by agriffard** [comment]
Renamed to MultiText.

Did you intend to use the direction setting for the checkboxes? (Vertical/Horizontal)

**Comment by deanmarcussen** [comment]
> Did you intend to use the direction setting for the checkboxes? (Vertical/Horizontal)

I was not, I reconnected it, because it wasn't registered. But then I saw it's not implemented. I would not be bothered with it. 
The vuemultiselect tags editor would provide a better option for reducing the amount of screen real estate this takes.

You don't like the vuemultiselect for this?

**Pull request merged**